### PR TITLE
[JENKINS-70141] Web component checkbox

### PR DESCRIPTION
**Pull request by scherler**
See [JENKINS-70141](https://issues.jenkins.io/browse/JENKINS-70141).
https://issues.jenkins.io/browse/JENKINS-70141
 
removed all checkbox code (css and jelly inner code) and moved it to a web component on its own. The functionality and styling should not have changed, if you see changes you found a bug. I created a new package based on yarn2 like the war one and further used lit to create the checkbox web component.  If you publish the package you can import the definition in any given npm driven project and use `<jenkins-checkbox />` as you would have (and still can in jelly interfaces) `<f:checkbox />`

### Proposed changelog entries

- Entry 1: Issue, human-readable text
- removed all checkbox code (css and jelly inner code) and moved it to a web component on its own.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7442"></a>

### CODE REVIEW

File: core/src/main/resources/lib/form/checkbox.jelly [hunk 1]
```diff
@@ -74,18 +74,20 @@ THE SOFTWARE.
...
+      title="${attrs.title}"
+      name="${name}"
+      value="${attrs.value}"
+      disabled="${readOnlyMode ? 'true' : null}"
+      checked="${value ? 'true' : null}"
+      id="${attrs.id}"
+      onclick="${attrs.readonly=='true' ? 'return false;' : attrs.onclick}"
+      class="${attrs.class} ${attrs.negative!=null ? 'negative' : null} ${attrs.checkUrl!=null?'validated':''}"
+      checkUrl="${attrs.checkUrl}" checkDependsOn="${attrs.checkDependsOn}" json="${attrs.json}"
+    ></jenkins-checkbox>
```
Comment (jglick) [inline core/src/main/resources/lib/form/checkbox.jelly:90]:
```suggestion
    />
```
conventionally

Comment (halkeye) [reply core/src/main/resources/lib/form/checkbox.jelly:90]: web components must have closed fully, no self closing.

Comment (jglick) [reply core/src/main/resources/lib/form/checkbox.jelly:90]: :thinking: how can it matter? This is a Jelly XML file.

Comment (halkeye) [reply core/src/main/resources/lib/form/checkbox.jelly:90]: i don't know enough about jelly. Does jelly automatically add the closing tag? or does it output it as is to the browser? if the second, then it needs to be closed in jelly too.

https://stackoverflow.com/questions/23961178/do-custom-elements-require-a-close-tag

Essentially only legacy tags are self closing (img, input, etc)

Comment (jglick) [reply core/src/main/resources/lib/form/checkbox.jelly:90]: > does it output it as is to the browser?

No, it parses XML, then generates (X)HTML. The input syntax should be irrelevant.

Comment (scherler) [reply core/src/main/resources/lib/form/checkbox.jelly:90]: The problem is the web component specification as pointed out by  @halkeye which does not allow self-closing, even if they do not have children.

Comment (jglick) [reply core/src/main/resources/lib/form/checkbox.jelly:90]: But again this source file is not a literal template; it is an XML document which is parsed. You can see this in the following passing test in a trivial plugin:

```java
package demo;
import hudson.Extension;
import hudson.model.InvisibleAction;
import hudson.model.RootAction;
@Extension public class X extends InvisibleAction implements RootAction {
    @Override public String getUrlName() {
        return "xxx";
    }
}
```

`X/index.jelly`:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?jelly escape-by-default='true'?>
<j:jelly xmlns:j="jelly:core" xmlns:l="/lib/layout">
      <l:layout>
          <l:main-panel>
              <xxx/>
          </l:main-panel>
      </l:layout>
</j:jelly>
```

```java
package demo;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.containsString;
import org.junit.Rule;
import org.junit.Test;
import org.jvnet.hudson.test.JenkinsRule;
public class XTest {
    @Rule public JenkinsRule r = new JenkinsRule();
    @Test public void tag() throws Exception {
        assertThat(r.createWebClient().goTo("xxx/").getWebResponse().getContentAsString(), containsString("<xxx></xxx>"));
    }
}
```

**Pull request closed without merging**